5.analyze-data/utils/well_significance_testing.py

```diff
@@ -0,0 +1,177 @@
+        The comparison_metric key contains the name of the comparison metric computed as a string.
+
+    _treatdf: pandas.Dataframe
+        The predicted probabilities and associated metadata for each treated cell (not in a control group).
+
+    _negcondf: pandas.Dataframe
+        The predicted probabilities and associated metadata for each cell in the negative control group.
+
+    _phenotype_cols: List
+        The names of the phenotype columns in the _treatdf and _negcondf dataframes.
+
+    _filt_cols: List
+        The names of the columns to group the treatment cells by before comparing the probabilities.
+
+    _control_cutoff: Integer
+        (Optional default=50) The minimum number of cells required for a negative control well to be included in the comparison.
+
+    _treat_cutoff: Integer
+        (Optional default=50) The minimum number of cells required for a treatment well (excluding negative control wells) to be included in the comparison.
+
+    Returns
+    -------
+    treatments: Dictionary
+        Contains the analysis information corresponding to each treatment
+    """
+
+    treatments = defaultdict(list)
+
+    # Iterate through each group
+    for filt_col_vals, group_treatdf in _treatdf.groupby(_filt_cols):
+
+        # The columns for keeping track of metadata and filtering the negative control cells
+        ref_cols = dict(zip(_filt_cols, filt_col_vals))
+
+        # The negative control cells
+        group_negdf = _negcondf.loc[(_negcondf["Metadata_Plate"] == ref_cols["Metadata_Plate"]) &
+                                   (_negcondf["Metadata_Model_Type"] == ref_cols["Metadata_Model_Type"])
+                                    ]
+
+        # Remove wells if the cell count is below the corresponding threshold
+        group_treatdf = filter_wells_by_cell_count(group_treatdf, _treat_cutoff)
+        group_negdf = filter_wells_by_cell_count(group_negdf, _control_cutoff)
+
+        # Compute the number of cells for each group
+        treat_cell_count = len(group_treatdf)
+        negcon_cell_count = len(group_negdf)
+        min_cell_count = min(treat_cell_count, negcon_cell_count)
+
+        # If there are no probability values that match the given well for some reason analyze the next treatment
+        if (min_cell_count > 0):
+
+            # Sample the treatment dataframe if the cell count for the treatments is larger than for the controls
+            if treat_cell_count > negcon_cell_count:
+                samp_treat = group_treatdf.sample(n=min_cell_count, random_state=0)
+                samp_neg = group_negdf
+
+            # Otherwise, keep all of the cells of the treatment group
+            # Stratify sample the negative control cells by the proportion of cells in each well
+            else:
+                samp_treat = group_treatdf
+                samp_neg = strat_samp_wells(group_negdf, min_cell_count)
+
+
```

Comment:
```suggestion
```
Recommend removing this extra line to match the format of the rest of the python script. Also recommend confirming that `black` has been performed to format this file and in the notebook.

Reply:
Removed